src/coreclr/debug/createdump/createdumppal.cpp

```diff
@@ -0,0 +1,289 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "createdump.h"
+
+#define INITGUID
+#include <guiddef.h>
+
+DEFINE_GUID(IID_IUnknown, 0x00000000, 0x0000, 0x0000, 0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46);
+
+#if defined(__linux__)
+#define PAL_FUNCTION_PREFIX "DAC_"
+#else
+#define PAL_FUNCTION_PREFIX
+#endif
+
+typedef int (*PFN_PAL_InitializeDLL)();
+typedef void (*PFN_PAL_TerminateEx)(int);
+
+typedef BOOL (*PFN_PAL_VirtualUnwindOutOfProc)(
+    CONTEXT *context,
+    KNONVOLATILE_CONTEXT_POINTERS *contextPointers,
+    PULONG64 functionStart,
+    SIZE_T baseAddress,
+    UnwindReadMemoryCallback readMemoryCallback);
+
+typedef BOOL (*PFN_PAL_GetUnwindInfoSize)(
+    SIZE_T baseAddress,
+    ULONG64 ehFrameHdrAddr,
+    UnwindReadMemoryCallback readMemoryCallback,
+    PULONG64 ehFrameStart,
+    PULONG64 ehFrameSize);
+
+typedef int (*PFN_WideCharToMultiByte)(
+    IN UINT codePage,
+    IN DWORD dwFlags,
+    IN LPCWSTR lpWideCharStr,
+    IN int cchWideChar,
+    OUT LPSTR lpMultiByteStr,
+    IN int cbMultiByte,
+    IN LPCSTR lpDefaultChar,
+    OUT LPBOOL lpUsedDefaultChar);
+
+bool g_initialized = false;
+PFN_PAL_InitializeDLL g_PAL_InitializeDLL = nullptr;
+PFN_PAL_TerminateEx g_PAL_TerminateEx = nullptr;
+PFN_PAL_VirtualUnwindOutOfProc g_PAL_VirtualUnwindOutOfProc = nullptr;
+PFN_PAL_GetUnwindInfoSize g_PAL_GetUnwindInfoSize = nullptr;
+PFN_WideCharToMultiByte g_WideCharToMultiByte = nullptr;
```

Comment:
For `WideCharToMultiByte`, can we use the APIs in `src/native/minipal/utf8.h` and `src/native/minipal/utf8.c` instead of the DAC's PAL? The PAL's implementation of `WideCharToMultiByte` is just a wrapper around that now.

Reply:
`#include <minipal/utf8.h>` should be globally accessible (it's in the shared cmake configs), just need to compile `utf8.c` once per-project.

Reply:
Is there a minipal function somewhere (I couldn't find it) get the length of a utf16 string.  These utf8 conversion function need it. The PAL WideCharToMultiByte impl uses PAL_wcslen to get it which would mean I just replace dyn-linking one PAL function for another.

The new createdump code that used to use WideCharToMultiByte  (so far):

```C++
std::string
ConvertString(const WCHAR* str)
{
    if (str == nullptr)
        return { };

    size_t cch = PAL_wcslen(str) + 1;
    int len = minipal_get_length_utf16_to_utf8((CHAR16_T*)str, cch, 0);
    if (len == 0)
        return { };

    ArrayHolder<char> buffer = new char[len + 1];
    minipal_convert_utf16_to_utf8((CHAR16_T*)str, cch, buffer, len + 1, 0);
    return std::string { buffer };
}
```

Reply:
There's https://github.com/dotnet/runtime/blob/52a3995f8692e5cd6dac4c033a9e73e320abc977/src/coreclr/minipal/Unix/dn-u16.cpp#L9

Reply:
The coreclrminipal library is dependent on these PAL functions: PAL_wcstoul, PAL__wcstoui64, PAL_wcstod so I just copied ut16_strlen source to my mini pal source and not linked that library.

Reply:
There are "independent" examples in mono: https://github.com/dotnet/runtime/blob/73638d6bad697a901ebea1891857a9f7037be3ab/src/mono/mono/eglib/giconv.c#L512-L518
(which ends up calling the same minipal conversion APIs)

First we need to call `minipal_get_length_utf16_to_utf8` to find out the required length, then allocate the buffer of that size, and call `minipal_convert_utf16_to_utf8` with the preallocated buffer.

Reply:
Ah, I see you have already included `dn-u16.h`. 👍